### ISSUE TITLE
Why are normal ajax calls exclude from the error_handler?

### ISSUE THREAD

**Issue by tonivdv** [nette/tracy#225]
Hi,

I'm using this library through https://github.com/webino/WebinoDebug . Previously we were using https://github.com/webino/ZF2NetteDebug where everything worked +- good. 

Anyway, we are currently upgrading step by step from zend 2.5 to 3, which means we have a lot of deprecation errors. Using tracy we can mute those deprecation errors by settings the severity to non-strict, however for are legacy ajax calls , the error handling of tracy is not "supported". Meaning we get the deprecation errors in the ajax calls.

So my questions are:

1. Why skip the error handling for ajax calls?

2. How can we avoid skipping them (it works great if I temporarily hack the isHtmlMode function)

3. Any other way to use the tracy error in ajax calls?

Thanks in advance

**Comment by tonivdv** [comment]
Hi,

It seems a good night sleep helps always. I was able to counter this by setting "error_reporting(E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED & ~E_NOTICE);" at the end of the bootstrap section of Zend.

That said, even if my problem is somehow fixed, I'm still eager to understand why tracy skips it's own error handler for ajax requests.

Cheers

**Comment by JanTvrdik** [comment]
>  I'm still eager to understand why tracy skips it's own error handler for ajax requests.

Why do you think that is the case? I'm not aware of any *skipping* for ajax request. Could you elaborate on what exactly do you mean by *skipping*?

---

Do you use Tracy 2.3 or Tracy 2.4? There are major differences in handling AJAX requests.

**Comment by tonivdv** [comment]
Hello @JanTvrdik ,

I'm speaking about this line https://github.com/nette/tracy/blob/master/src/Tracy/Debugger.php#L423

Excerpt 

```php
    } else {
		self::fireLog(new ErrorException($message, 0, $severity, $file, $line));
		return Helpers::isHtmlMode() || Helpers::isAjax() ? NULL : FALSE; // FALSE calls normal error handler
	}
```

Excerpt of the Helper code:

```php
    /** @internal */
	public static function isHtmlMode()
	{
		return empty($_SERVER['HTTP_X_REQUESTED_WITH']) && empty($_SERVER['HTTP_X_TRACY_AJAX'])
			&& PHP_SAPI !== 'cli'
			&& !preg_match('#^Content-Type: (?!text/html)#im', implode("\n", headers_list()));
	}

	/** @internal */
	public static function isAjax()
	{
		return isset($_SERVER['HTTP_X_TRACY_AJAX']) && preg_match('#^\w{10}\z#', $_SERVER['HTTP_X_TRACY_AJAX']);
	}
```

As you can, when handling normal ajax calls (no X_TRACY_AJAX) it returns false, which returns the error handler to the default one.

The issue is the same in tracy 2.3 (in that version there is simply no check for tracy ajax calls).

Hope this clarifies my question :)

Thanks

**Comment by JanTvrdik** [comment]
@tonivdv And what you expect Tracy to do?

The last branch (`return Helpers::isHtmlMode() || Helpers::isAjax() ? ...`) is executed when Tracy in debug mode receives information about error but Bluescreen cannot be rendered (because strictMode is not enabled). It checks for HTML or X_TRACY_AJAX mode (meaning debug bar will be visible) nad if neither of them is active Tracy is out of options how to inform developer about the error, so it fallbacks to PHP error handler.

If strict mode was enabled (which is recommended as best practise), Tracy would create Bluescreen and write it to log directory.

**Comment by tonivdv** [comment]
Hello @JanTvrdik ,

The issue, for us, is that we are telling Tracy (through the zf module) that we don't want to be in strict mode (hence we need to disable the E_DEPRECATED and E_USER_DEPRECATED because we cannot solve all deprecation errors at once ). And this works fine, until we call some code through AJAX, then Tracy will fallback to the default error_handler (which is not per se an issue), but during that phase it also put back the strict mode on, or changed the error_reporting mode.

In short, we don't want Tracy to change the error_reporting level if it's ajax! 

Note that we are not using the full potential of Tracy (debug bar etc) as this is not supported (yet) in the ZF module we are using.

And I completely agree, strict mode is the way to go, but in big projects with legacy code you cannot simply stop all dev to fix many deprecations. We do them step by step, but therefor we need to be able to ignore the deprecation levels.

Cheers

**Comment by JanTvrdik** [comment]
> we don't want to be in strict mode 

Just to clarify, Tracy's "strictMode" (`Tracy\Debugger::$strictMode`) means stopping execution in situations where PHP would normally continue (notice, warnings...). But it still respects error reporting level – if you configure PHP to ignore deprecated errors, they will remain ignored even when strict mode is enabled.

> but during that phase it also put back the strict mode on, or changed the error_reporting mode.

This seems odd, there are only two situactions where Tracy changes error reporting level

1) in [Tracy\Debugger::enable()](https://github.com/nette/tracy/blob/v2.4.2/src/Tracy/Debugger.php#L177) when initially started
2) in [Tracy\Debugger::errorHandler()](https://github.com/nette/tracy/blob/v2.4.2/src/Tracy/Debugger.php#L369) when scream mode (`Tracy\Debugger::$streamMode)` is on

However scream mode should only be enabled when debugging some very rare edge cases. It must not be normally enabled because it breaks normal PHP behavior.

**Comment by tonivdv** [comment]
@JanTvrdik Thanks. I will check this tomorrow, because if I remember correctly the screamMode was called. If you say it shouldn't in normal cases or without doing anything specific it could be related to a config in the ZF module which encaps this library.

**Comment by tonivdv** [comment]
Hello @JanTvrdik ,

It's not the scream mode as this is not called, but it's the error_reporting set in [Tracy\Debugger::enable()](https://github.com/nette/tracy/blob/v2.4.2/src/Tracy/Debugger.php#L177) which sets it to E_ALL. When I comment this my Ajax calls work like a charm.

In order to test this outside our complex application, I quickly created a Silex application to pinpoint the issue. You can find it at https://github.com/tonivdv/tracyAjaxExample

Hope this will help you better understand what I mean.

Cheers

**Comment by JanTvrdik** [comment]
If that is the case, just call `error_reporting()` after `Tracy\Debugger::enable()` is called and everything should work fine.

**Comment by tonivdv** [comment]
Hello @JanTvrdik ,

But wouldn't it be better to do that within the library? Currently I'm fixing it like this, but I would expect the library to "fix" this for me?!

**Comment by JanTvrdik** [comment]
To do what in the library? To NOT call `error_reporting()` in `Tracy\Debugger::enable()`? 

`Tracy\Debugger::enable()` is meant to be called very, very early (ideally it should be the first thing you call after requiring Composer autoloader, although that is not always possible) and it intentionally normalizes PHP configuration to safe values and even [stops the execution altogether](https://api.nette.org/2.4/source-Tracy.Debugger.php.html#175) when the environment is not and cannot be made secure.

**Comment by tonivdv** [comment]
@JanTvrdik Just to be sure we are understanding each other correctly :)

From this library point of view, the behaviour explained in the example at https://github.com/tonivdv/tracyAjaxExample is completely normal?

Because imho it's not. For me, without changing any code, the error level should not change whether it comes from a normal http request or an ajax http request.

**Comment by JanTvrdik** [comment]
> For me, without changing any code, the error level should not change whether it comes from a normal http request or an ajax http request.

But that's not what is happening. The error level is always the same regardless of whether the request comes from AJAX or not. The difference is that in the normal request the error is shown in debug bar, however in AJAX request there is no debug bar (*), so Tracy has no way of showing you the error and therefore fallbacks to default PHP error handler which simply prints the error.

 * Since Tracy 2.4, debug bar for AJAX request is supported as well if you either start sessions before calling `Debugger::enable()` or if you call manually `Debugger::dispatch()` later. In the specific case of your example you also need to comment out `$('body').html(data);` as it removes the main debug bar to which the AJAX bar should be appended to.

![image](https://cloud.githubusercontent.com/assets/175109/18839256/9c55aa56-840a-11e6-8e12-40b5901d73aa.png)

**Comment by tonivdv** [comment]
Hello @JanTvrdik ,

Thank you for explaining a bit more about it. I started reading a bit more the code to understand how it works behind (not always have the time to do that sorry) and it's clear. I think the main confusion comes from the zf module which imho does not do a very good job at incorporating this library.

Anyway I can live with the workaround until a better zf module exists or the existing one is improved.

Thanks again for your time.

Cheers
